ocaml/xapi-cli-server/cli_operations.ml

```diff
@@ -5462,17 +5462,49 @@ let wait_for_task_complete rpc session_id task_id =
     Thread.delay 1.0
   done
 
+let check_task_status ?(quiet_on_success = false) ~rpc ~session_id ~task ~fd
+    ~label ~ok () =
+  (* if the client thinks it's ok, check that the server does too *)
+  match Client.Task.get_status ~rpc ~session_id ~self:task with
+  | `success when ok ->
+      if not quiet_on_success then
+        marshal fd (Command (Print (Printf.sprintf "%s succeeded" label)))
+  | `success ->
+      marshal fd
+        (Command
+           (PrintStderr (Printf.sprintf "%s failed, unknown error.\n" label))
+        ) ;
+      raise (ExitWithError 1)
+  | `failure ->
+      let result = Client.Task.get_error_info ~rpc ~session_id ~self:task in
+      if result = [] then
+        marshal fd
+          (Command
+             (PrintStderr (Printf.sprintf "%s failed, unknown error\n" label))
+          )
+      else
+        raise (Api_errors.Server_error (List.hd result, List.tl result))
+  | `cancelled ->
+      marshal fd (Command (PrintStderr (Printf.sprintf "%s cancelled\n" label))) ;
+      raise (ExitWithError 1)
+  | _ ->
+      marshal fd (Command (PrintStderr "Internal error\n")) ;
+      (* should never happen *)
+      raise (ExitWithError 1)
+
 let download_file rpc session_id task fd filename uri label =
   marshal fd (Command (HttpGet (filename, uri))) ;
   let response = ref (Response Wait) in
   while !response = Response Wait do
     response := unmarshal fd
   done ;
   let ok =
-    match !response with
-    | Response OK ->
+    match (!response, filename <> "") with
+    | Response OK, true ->
         true
-    | Response Failed ->
+    | Response OK, false ->
+        false
```

Comment:
Seems when filename is "", it will print to `stdout`: https://github.com/xapi-project/xen-api/blob/8ab1fe3340f79299204939ed7b2a1f6cde8e3d7a/ocaml/xe-cli/newcli.ml#L726-L727, so it is also supported then it is not a failure?

Reply:
The code `newcli.ml` runs as `xe` client, and the code in `cli_operations.ml` runs in xapi as server.
In this case, the xapi receives the request from `xe` client, send `HttpGet` to the `xe` client, and then
1. the `xe` client opens `stdout` indeed,
2. but the xapi writes error to client and raise exception eventually.
 No behaviour being changed here. The xapi (server) determines if it is an error or not.

Reply:
Thanks for the explanation!
`No behaviour being changed here.`, I think there is some problem with the original code: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi-cli-server/cli_operations.ml#L5490-L5491, seems there was no handling for the case: `ok && filename = ""`?